File: runtime/lua/modules/btea/models/list.go
package models

/*
I'll break down what we need in the list binding implementation to bridge between Go and Lua.

The key components we'll need are:

1. **Main List Struct & Registration**
```go
// List wraps list.Model for Lua
type List struct {
    model list.Model
    // We'll need to store delegate and items with Lua references
    delegate *lua.LFunction
    items    []*lua.LUserData
    // Store Lua state for calling back into Lua
    luaState *lua.LState
}

// RegisterList registers list to Lua state
func RegisterList(L *lua.LState, mod *lua.LTable)
```

2. **Item Interface Bridge**
- Need wrapper for Lua tables/userdata to implement list.Item
- Must handle calling Lua methods (FilterValue, Title, Description)
- Cache results to avoid excessive Lua calls

3. **Delegate Bridge**
- Wrapper to make Lua tables act as list.ItemDelegate
- Handle Height, Spacing, Render, Update methods
- Convert between Go and Lua types for messages and commands
- Manage writer for Render method

4. **Constructor & Config**
```go
// Config needs to handle:
- Items array
- Delegate
- Width/Height
- All display flags (showTitle, etc)
- Styles
- Key bindings
- Status bar customization
```

5. **Methods to Expose to Lua**
```go
// Core methods
- Update (handle msgs)
- View (return string)

// Item management
- Items
- SetItems
- SetItem
- InsertItem
- RemoveItem
- SelectedItem

// Navigation
- CursorUp
- CursorDown
- PrevPage
- NextPage
- Select
- ResetSelected

// Filtering
- FilterState
- FilterValue
- SettingFilter
- IsFiltered
- ResetFilter

// Display control
- SetWidth
- SetHeight
- SetShowTitle
- SetShowFilter
- SetShowStatusBar
- SetShowPagination
- SetShowHelp
- SetStatusBarItemName

// Spinner
- StartSpinner
- StopSpinner
- ToggleSpinner

// Status
- NewStatusMessage
```

6. **Message/Command Conversion**
- Convert tea.Msg to Lua
- Convert commands back to Lua callbacks
- Handle key events
- Handle window size events
- Handle filter matches

7. **Style Integration**
- Convert between Lua style tables and lipgloss styles
- Handle all list-specific styles
- Manage style inheritance and updates

8. **Key Binding Integration**
- Convert between Lua key binding tables and bubbles/key
- Manage help text
- Handle key mapping updates

9. **Special Considerations**
- Memory management of Lua references
- Error handling for Lua callbacks
- Performance optimization for frequent operations
- Proper cleanup when list is garbage collected
- Thread safety for Lua state access

10. **Documentation & Examples**
- GoDoc comments explaining each component
- Usage examples
- Common pitfalls to avoid
- Performance considerations

Would you like me to elaborate on any of these components or discuss specific challenges in implementing them?
*/

File: app/terminal/btea/list.lua
local bapp = require("bapp")

function App()
    local app = bapp.new()

    local items = {
        {
            title = "Buy groceries",
            description = "Milk, eggs, bread, cheese"
        },
        {
            title = "Finish report",
            description = "Due by end of day Friday"
        }
    }

    local delegate = {
        height = function() return 1 end,
        spacing = function() return 1 end,
        render = function(m, index, item)
            return item.title .. " - " .. item.description
        end
    }

    app.list = btea.new_list({
        width = app.window.width,
        height = app.window.height - 2,
        items = items,
        delegate = delegate
    })

    -- Setup key bindings
    app.keys = bapp.create_keys({
        quit = { keys = { "q", "ctrl+c" } },
        up = { keys = { "up", "k" } },
        down = { keys = { "down", "j" } }
    })

    local function update(self, msg)
        if msg.key then
            if self.keys.quit:matches(msg) then
                return true
            elseif self.keys.up:matches(msg) then
                self:dispatch(self.list:cursor_up())
            elseif self.keys.down:matches(msg) then
                self:dispatch(self.list:cursor_down())
            end

            self:dispatch(self.list:update(msg))
        end
        return false
    end

    local function view(self)
        return self.list:view()
    end

    app:run(update, view)

    return app
end

return App
